### PR TITLE
feat: TypeScript 4.2 syntax support

### LINKED ISSUE #2935: TypeScript 4.2 Syntax Support
https://devblogs.microsoft.com/typescript/announcing-typescript-4-2-beta/

This issue is just to track all of the new features and their implementation state in this project.
As with all releases, we will not necessarily to support all features until closer to the full release when everything the features are stabilised.

*Please be patient.*

---

# Leading/Middle Rest Elements in Tuple Types

```ts
let foo: [...string[], number];

foo = [123];
foo = ["hello", 123];
foo = ["hello!", "hello!", "hello!", 123];

let bar: [boolean, ...string[], boolean];

bar = [true, false];
bar = [true, "some text", false];
bar = [true, "some", "separated", "text", false];
```

This will not require any AST changes.

# Smarter Type Alias Preservation

This will not require any AST changes.

Having a quick look on [ts-ast-viewer](https://ts-ast-viewer.com/#code/KYDwDg9gTgLgBDAnmYcBCBDAzgSwMYAKUOAtjjDgG6oC8cAdgK4kBGwUcAPnFjMfQHMucFhAgAbYBnoBuAFBzQkWHABmjengoR6cACYQAyjEarVACkoZxjYAC502fEVLkqwAJQPMuQsTIU1MIaesCqOPTAenAA3nJwCQk4qnDmALIYMAAWAHRQ0gYk5h5wADxwAAw5AKwlcYkNCVDAJlC6IWERUfKNAL4Kjc2tulY2wD29QA), it looks like the types API continues to work as it did previously, but there is additional information stored on the type (the internal `origin` property) which stores the de-normalised union.

I don't think this will impact any of our rules, but we'll find out with tests.

# Template Literal Expressions Have Template Literal Types

This will not require any AST changes.

Having a quick look on [ts-ast-viewer](https://ts-ast-viewer.com/#code/MYewdgzgLgBGBccCuBbARgUwE4wLwwEYAmAZgG4AoCgempgAkBDCGKACw1YE8AHTgAwAkAbzCpMWAL48AHvwqhIsCATwwho6XLIwgA), it just means the types will be template literal types.

I don't think we've gone through and fully checked compat with our rules and template literal types yet - but given this will increase the prevalence of template literal types, we might have to go through and check our rules.

# `abstract` Construct Signatures

```ts
let Ctor: abstract new () => HasArea = Shape;
```

This ***will*** require AST changes.

# `--noPropertyAccessFromIndexSignature`

This will not require any AST changes.

The `dot-notation` extension rule will require work so that it can coexist with this compiler option (#3104).

# Other changes with no impact to us

- Stricter Checks for the in Operator (The `in` Operator No Longer Allows Primitive Types on the Right Side)
  - Once we eventually have syntax errors in our parser (#1852) we could expose this error
- `--explainFiles` to understand why files have been included in a program
- Relaxed Rules Between Optional Properties and String Index Signatures
- Declare Missing Helper Function
- `noImplicitAny` Errors Apply to Loose `yield` Expressions
- Type Arguments in JavaScript Are Not Parsed as Type Arguments
- TypeScript’s lift Callback in visitNode Uses a Different Type

### PR DESCRIPTION
**Pull request by armano2**
### Ast structure addition:
- `JSXNamespacedName` will now be generated when jsx namespace is used in jsx attribute or tag name
  ```jsx
  <foo:bar />;
  ```
  ```js
  JSXNamespacedName <: Node {
    type: "JSXNamespacedName"
    namespace: JSXIdentifier
    name: JSXIdentifier
  }
  ```
- `TSConstructorType` will now support abstract construct signatures 
  ```js
  TSConstructorType <: Node {
    abstract: boolean
  }
  ```
  ```ts
  let Ctor: abstract new () => HasArea = Shape;
  ```

### TODO:
- [x] update typescript to 4.2.x
- [x] add support for abstract construct signatures 
- [x] add parsing of `JSXNamespacedName`
- [x] investigate `<a:b.c></a:b.c>` and  `<this:b></this:b>`  https://github.com/microsoft/TypeScript/issues/42963
- [ ] investigate usage of `noPropertyAccessFromIndexSignature `

fixes #2935

### CODE REVIEW

File: packages/types/src/ts-estree.ts [hunk 1]
```diff
@@ -374,6 +374,7 @@ export type Expression =
   | SequenceExpression
   | SpreadElement
   | TSAsExpression
+  | TSTypeAssertion
```
Comment (armano2) [inline packages/types/src/ts-estree.ts:377]: this change is unrelated to upgrade of typescript

File: packages/typescript-estree/src/ts-estree/estree-to-ts-node-types.ts [hunk 2]
```diff
@@ -157,15 +157,11 @@ export interface EstreeToTsNodeTypes {
     | ts.ConstructorDeclaration;
   [AST_NODE_TYPES.TSArrayType]: ts.ArrayTypeNode;
   [AST_NODE_TYPES.TSAsExpression]: ts.AsExpression;
-  [AST_NODE_TYPES.TSCallSignatureDeclaration]: ts.PropertySignature;
+  [AST_NODE_TYPES.TSCallSignatureDeclaration]: ts.CallSignatureDeclaration;
```
Comment (armano2) [inline packages/typescript-estree/src/ts-estree/estree-to-ts-node-types.ts:160]: this change is unrelated to upgrade of typescript

File: packages/typescript-estree/src/ts-estree/estree-to-ts-node-types.ts [hunk 3]
```diff
@@ -157,15 +157,11 @@ export interface EstreeToTsNodeTypes {
...
   [AST_NODE_TYPES.TSClassImplements]: ts.ExpressionWithTypeArguments;
   [AST_NODE_TYPES.TSConditionalType]: ts.ConditionalTypeNode;
   [AST_NODE_TYPES.TSConstructorType]: ts.ConstructorTypeNode;
-  [AST_NODE_TYPES.TSConstructSignatureDeclaration]:
-    | ts.ConstructorTypeNode
-    | ts.FunctionTypeNode
-    | ts.ConstructSignatureDeclaration
-    | ts.CallSignatureDeclaration;
+  [AST_NODE_TYPES.TSConstructSignatureDeclaration]: ts.ConstructSignatureDeclaration;
```
Comment (armano2) [inline packages/typescript-estree/src/ts-estree/estree-to-ts-node-types.ts:164]: this change is unrelated to upgrade of typescript but compiler started complaining about this now

**Comment by alitaheri** [comment]
Will this change also support this syntax?

```ts
export type Type<T extends number> = `my-id-${T}`;
// Getting error:                                       ^ Parsing error: Type expected.
```

**Comment by bradzacher** [comment]
Template string types are already supported as part of our TS 4.1 support.

If you're having issues, ensure you update your versions.

**Comment by armano2** [comment]
update: `JSXNamespacedName` will now be created

File: packages/typescript-estree/src/convert.ts [hunk 4]
```diff
@@ -539,27 +579,14 @@ export class Converter {
...
+          property: this.convertJSXIdentifier(node.name),
         });
         break;
 
       case SyntaxKind.ThisKeyword:
-        result = this.createNode<TSESTree.JSXIdentifier>(node, {
-          type: AST_NODE_TYPES.JSXIdentifier,
-          name: 'this',
-        });
-        break;
```
Comment (armano2) [inline packages/typescript-estree/src/convert.ts:554]: i'm unsure how we want to parse this, as this is invalid

i followed lead of babel on this 
```
<this:x />
```

https://astexplorer.net/#/gist/270cbd05d523f900d08730adce14815d/4ed17997e784261e930cafe2188ff705a53786a2

**Review by bradzacher — APPROVED** [review]
thanks!

File: packages/typescript-estree/src/convert.ts [hunk 5]
```diff
@@ -2407,36 +2431,40 @@ export class Converter {
...
-          default:
-            type = AST_NODE_TYPES.TSConstructorType;
-            break;
-        }
+        const type =
+          node.kind === SyntaxKind.ConstructSignature
+            ? AST_NODE_TYPES.TSConstructSignatureDeclaration
+            : node.kind === SyntaxKind.CallSignature
+            ? AST_NODE_TYPES.TSCallSignatureDeclaration
+            : AST_NODE_TYPES.TSFunctionType;
```
Comment (bradzacher) [inline packages/typescript-estree/src/convert.ts:2459]: IMO nested ternaries are awful for readability.
If your intention is to make `type` a `const`, I'd suggest something like an IIFE.

```suggestion
        const type = ((): AST_NODE_TYPES => {
          switch (node.kind) {
            case SyntaxKind.ConstructSignature:
              return AST_NODE_TYPES.TSConstructSignatureDeclaration;
              
            case SyntaxKind.CallSignature:
              return AST_NODE_TYPES.TSCallSignatureDeclaration;
              
            case SyntaxKind.FunctionType:
              return AST_NODE_TYPES.TSFunctionType;
              
            case SyntaxKind.ConstructorType:
            default:
              return AST_NODE_TYPES.TSConstructorType;
          }
        })();
```

**Review by bradzacher — COMMENTED** [review]
needs to be updated with master

**Review by bradzacher — APPROVED** [review]

File: packages/typescript-estree/src/convert.ts (hunk 5 above)
Comment (armano2) [reply packages/typescript-estree/src/convert.ts:2459]: yes and no, bdw, nowthis case only handles:
`SyntaxKind.FunctionType`, `SyntaxKind.ConstructSignature`, `SyntaxKind.CallSignature` 

as `ConstructorType` is handled separately

Comment (armano2) [reply packages/typescript-estree/src/convert.ts:2459]: if you really want we can change it to
```suggestion
        let type:
          | AST_NODE_TYPES.TSConstructSignatureDeclaration
          | AST_NODE_TYPES.TSCallSignatureDeclaration
          | AST_NODE_TYPES.TSFunctionType;
        switch (node.kind) {
          case SyntaxKind.ConstructSignature:
            type = AST_NODE_TYPES.TSConstructSignatureDeclaration;
            break;
          case SyntaxKind.CallSignature:
            type = AST_NODE_TYPES.TSCallSignatureDeclaration;
            break;
          case SyntaxKind.FunctionType:
          default:
            type = AST_NODE_TYPES.TSFunctionType;
            break;
        }
```

Comment (bradzacher) [reply packages/typescript-estree/src/convert.ts:2459]: it was just a suggestion if you wanted to do it during the master merge - I'm happy to merge this as is.

**Comment by kentcdodds** [comment]
Hey friends 👋

I think this was supposed to be updated: https://github.com/typescript-eslint/typescript-eslint/blob/bf904ec72db57174fec531f61e9427230662553e/packages/typescript-estree/src/parser.ts#L28

**Comment by kentcdodds** [comment]
Made a PR: https://github.com/typescript-eslint/typescript-eslint/pull/3127

**Comment by DerZyklop** [comment]
@alitaheri is your issue solved?

**Comment by armano2** [comment]
@DerZyklop yes version has been updated in patch release 4.16.1

**Comment by alitaheri** [comment]
@DerZyklop yes, tested it just now. no more errors 👍🏼
